VSS.md

```diff
@@ -0,0 +1,518 @@
+
+**Monitoring CUJ** \- As a user, I need to be able to monitor indexes and re-indexing with simple observability metrics
+
+- Requirement: Ability to monitor index base statistics  
+- Requirement: Ability to monitor cross indexes, aggregated, view statistics.
+
+**Backfill CUJ** \- As a user, I need to be able to create an index after the keyspace contains vector entries
+
+- Requirement: Index creation triggers a backfill process which indexes the existing keyspace entries.
+
+**Hybrid Query CUJ** \- As a user, I want to be able to perform hybrid queries to incorporate filtering criteria in addition to best-match vector search
+
+- Requirement: Support for filtering numeric and tag-based filters.  
+- Requirement: Support for boolean combinations of operators in combination with filtering (AND, OR, NOT)
+
+**Result Set Management CUJ** \- As a user I want to dictate which fields are returned in a search
+
+- Requirement: Support query controls to set the desired returned fields.
+
+**ACLs CUJ** \- As a user of Access Control Lists (ACLs), I want to manage user-based permissions, specific command permissions, etc to manage vector search access
+
+- Requirement: Existing ACls functionality should extend to vector search commands
+
+**Persistence CUJ** \- As a user I want to be able to leverage RDB and/or AOF for persisting and loading indexes
+
+- Requirement: Index metadata should be persisted leveraging RDB  
+- Requirement: Index types that take significant time to rebuild, such as vector indexes, must support persistence of the index payload.
+
+**Tuning CUJ** \- As a user who is also managing the infrastructure, I want to tune my configurations, like number of threads, to optimize my search performance
+
+- Requirement: Configs like number of threads should be exposed and tunable to users
+
+## Current Specification
+
+The search module allows a user to create any number of indexes. Each index consists of any number of typed fields and covers a specified subset of the keys within a Valkey database.
+
+The creation of an index initiates a backfill process in the background that will scan the keyspace and insert all covered keys into the newly created index. Mutations of a key are automatically reflected in the index. This process is automatic after the creation of the index. Mutation of keys is explicitly allowed during the backfill process. The current state of a backfill may be interrogated by a client using the FT.INFO command.
+
+A query operation is a two step process. In the first step, a query is executed that generates a set of keys and distances. In the second step, that list of keys and distances is used to generate a response.
+
+The first step is performed in a background thread, while the second step is performed by the mainthread. If a key produced by step 1 is mutated before step 2 is executed by the mainthread, the result of the query may not be consistent with the latest state of the keys.
+
```

Comment:
Can we tolerance this inconsistent?

Reply:
In CMD mode, the application could use LUA or multi/exec to avoid this issue-- with the concurrent reduction in throughput.

For CME, the question is more interesting as there is no attempt to synchronize the per-shard queries. Thus while it may be possible to avoid this issue in the originating shard (i.e., do something like the LUA multi/exec described above), the application will still have to deal with the fact that the returned results might be out-of-date for all of the non-originating shards.